# <center>Python Web Scraping [27 exercises with solution]</center>

### Web Scrapping:

- Web scraping or web data extraction is data scraping used for extracting data from websites. 
- Web scraping softwares are used to access the World Wide Web directly using the Hypertext Transfer Protocol, or through a web browser. 
- While web scraping can be done manually by a software user, the term typically refers to automated processes implemented using a bot or web crawler. 
- It is a form of copying, in which specific data is gathered and copied from the web, typically into a central local database or spreadsheet, for later retrieval or analysis.

### Python request module :

- Requests allows user to send organic, grass-fed HTTP/1.1 requests, without the need for manual labor. 
- There’s no need to manually add query strings to your URLs, or to form-encode your POST data.

### 1. Write a Python program to test if a given page is found or not on the server.

In [1]:
from urllib.request import urlopen
from urllib.error import HTTPError
from urllib.error import URLError

try:
    html = urlopen('https://abcxyz.com')
except HTTPError as e:
    print('HTTP error!')
except URLError as e:
    print('URL error! Server not found.')
else:
    print(html.read())
    
print('\n')

try:
    html = urlopen('http://www.example.com/')
    print(html.read())
except HTTPError as e:
    print('HTTP error!')
except URLError as e:
    print('URL error! Server not found.')

URL error! Server not found.


b'<!doctype html>\n<html>\n<head>\n    <title>Example Domain</title>\n\n    <meta charset="utf-8" />\n    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />\n    <meta name="viewport" content="width=device-width, initial-scale=1" />\n    <style type="text/css">\n    body {\n        background-color: #f0f0f2;\n        margin: 0;\n        padding: 0;\n        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;\n        \n    }\n    div {\n        width: 600px;\n        margin: 5em auto;\n        padding: 2em;\n        background-color: #fdfdff;\n        border-radius: 0.5em;\n        box-shadow: 2px 3px 7px 2px rgba(0,0,0,0.02);\n    }\n    a:link, a:visited {\n        color: #38488f;\n        text-decoration: none;\n    }\n    @media (max-width: 700px) {\n        div {\n            margin: 0 auto;\n            width: auto;\n        }\n    }\n    </style>    \

### 2. Write a Python program to download and display the content of robot.txt for en.wikipedia.org.

In [3]:
import requests
response = requests.get('https://en.wikipedia.org/robots.txt')
print(response.text)

# robots.txt for http://www.wikipedia.org/ and friends
#
# Please note: There are a lot of pages on this site, and there are
# some misbehaved spiders out there that go _way_ too fast. If you're
# irresponsible, your access to the site may be blocked.
#

# Observed spamming large amounts of https://en.wikipedia.org/?curid=NNNNNN
# and ignoring 429 ratelimit responses, claims to respect robots:
# http://mj12bot.com/
User-agent: MJ12bot
Disallow: /

# advertising-related bots:
User-agent: Mediapartners-Google*
Disallow: /

# Wikipedia work bots:
User-agent: IsraBot
Disallow:

User-agent: Orthogaffe
Disallow:

# Crawlers that are kind enough to obey, but which we'd rather not have
# unless they're feeding search engines.
User-agent: UbiCrawler
Disallow: /

User-agent: DOC
Disallow: /

User-agent: Zao
Disallow: /

# Some bots are known to be trouble, particularly those designed to copy
# entire sites. Please obey robots.txt.
User-agent: sitecheck.internetseer.com
Disallow: /

User-agent: 

### 3. Write a Python program to get the number of datasets currently listed on data.gov.

Module `cssselect` don't come with python as built in. I installed it using - `conda install cssselect`.

**lxml** - the most feature-rich and easy-to-use library for processing XML and HTML in the Python language.

In [25]:
from lxml import html     
import requests

response = requests.get('http://www.data.gov/')               # Output: <Response [200]
doc_gov = html.fromstring(response.text)                      # Output: Element html at 0x22d8aed09a8
link_gov = doc_gov.cssselect('small a')[0]                    # Runs the CSS expression on this element and its children.
print('Number of datasets currently listed on \'data.gov\' website is: ')   # Don't know what 'small a' means. I think it denotes small and anchor tag of html.
print(link_gov.text)

Number of datasets currently listed on 'data.gov' website is: 
250,060 datasets


### 4. Write a Python program to convert an address (like "1600 Amphitheatre Parkway, Mountain View, CA") into geographic coordinates (like latitude 37.423021 and longitude -122.083739).

**Geocodin:** Geocoding is the process of converting addresses (like "1600 Amphitheatre Parkway, Mountain View, CA") into geographic coordinates (like latitude 37.423021 and longitude -122.083739), which you can use to place markers on a map, or position the map.

To geolocate a single address, we can use Geopy python library. Geopy has different Geocoding services that we can choose from, including Google Maps, ArcGIS, AzureMaps, Bing, etc. Some of them require API keys, while others do not need.

In [40]:
# Will solve later. Facing some problem with API keys.

### 5. Write a Python program to display the name of the most recently added dataset on data.gov. 

In [53]:
from lxml import html
import requests

response = requests.get('https://catalog.data.gov/dataset?q=&sort=metadata_created+desc')
doc = html.fromstring(response.text)
title = doc.cssselect('h3.dataset-heading')[0].text_content()      # <h3 class="dataset-heading">
# Output of title: '\n        \n\n        \n\n        Observations of total phosphorous (TP) to support nearshore nutrient modeling, 2015.\n        \n        \n\n\n\n\n\n      '
# As we have lots of whitespace so we will use strip to remove them:

print("The name of the most recently added dataset on \'data.gov\' website:")
print(title.strip())

The name of the most recently added dataset on 'data.gov' website:
Observations of total phosphorous (TP) to support nearshore nutrient modeling, 2015.


### 6. Write a Python program to extract h1 tag from example.com.

In [6]:
from urllib.request import urlopen
from bs4 import BeautifulSoup

html = urlopen('http://www.example.com/')
bsh = BeautifulSoup(html.read(), 'html.parser')
print(bsh.h1)

<h1>Example Domain</h1>


### 7. Write a Python program to extract and display all the header tags from en.wikipedia.org/wiki/Main_Page. 

In [21]:
from urllib.request import urlopen
from bs4 import BeautifulSoup

html = urlopen('https://en.wikipedia.org/wiki/Main_Page')
bsh = BeautifulSoup(html.read(), 'html.parser')
headers = bsh.find_all(['h1', 'h2', 'h3', 'h4', 'h5', 'h6'])
print('List of all headers tag:')
print('=' * 25, '\n')
print(*headers, sep = '\n\n')

List of all headers tag:

<h1 class="firstHeading" id="firstHeading" lang="en">Main Page</h1>

<h2 id="mp-tfa-h2" style="margin:0.5em; background:#cef2e0; font-family:inherit; font-size:120%; font-weight:bold; border:1px solid #a3bfb1; color:#000; padding:0.2em 0.4em;"><span id="From_today.27s_featured_article"></span><span class="mw-headline" id="From_today's_featured_article">From today's featured article</span></h2>

<h2 id="mp-dyk-h2" style="clear:both; margin:0.5em; background:#cef2e0; font-family:inherit; font-size:120%; font-weight:bold; border:1px solid #a3bfb1; color:#000; padding:0.2em 0.4em;"><span class="mw-headline" id="Did_you_know...">Did you know...</span></h2>

<h2 id="mp-itn-h2" style="margin:0.5em; background:#cedff2; font-family:inherit; font-size:120%; font-weight:bold; border:1px solid #a3b0bf; color:#000; padding:0.2em 0.4em;"><span class="mw-headline" id="In_the_news">In the news</span></h2>

<h2 id="mp-otd-h2" style="clear:both; margin:0.5em; background:#cedff2

### 8. Write a Python program to extract and display all the image links from en.wikipedia.org/wiki/Peter_Jeffrey_(RAAF_officer). 

In [39]:
from urllib.request import urlopen
from bs4 import BeautifulSoup
import re

html = urlopen('https://en.wikipedia.org/wiki/Peter_Jeffrey_(RAAF_officer)')
bsh = BeautifulSoup(html.read(), 'html.parser')
images = bsh.find_all('img', {'src': re.compile('.jpg')})   #Compile a regular expression pattern, returning a Pattern object.
for image in images:
    print(image['src'], '\n')                               # I am not clear about how this print works. 

//upload.wikimedia.org/wikipedia/commons/thumb/a/af/NlaJeffrey1942-43.jpg/220px-NlaJeffrey1942-43.jpg 

//upload.wikimedia.org/wikipedia/commons/thumb/c/c5/008315JeffreyTurnbull1941.jpg/260px-008315JeffreyTurnbull1941.jpg 

//upload.wikimedia.org/wikipedia/commons/e/ea/021807CameronJeffrey1941.jpg 

//upload.wikimedia.org/wikipedia/commons/thumb/9/92/AC0072JeffreyTruscottKittyhawks1942.jpg/280px-AC0072JeffreyTruscottKittyhawks1942.jpg 

//upload.wikimedia.org/wikipedia/commons/thumb/2/26/VIC1689Jeffrey1945.jpg/280px-VIC1689Jeffrey1945.jpg 



### 9. Write a Python program to get 90 days of visits broken down by browser for all sites on data.gov.

### 10. Write a Python program to that retrieves an arbitary Wikipedia page of "Python" and creates a list of links on that page.

### 11. Write a Python program to check whether a page contains a title or not.

### 12. Write a Python program to list all language names and number of related articles in the order they appear in wikipedia.org.

### 13. Write a Python program to get the number of people visiting a U.S. government website right now.

*Source*: [https://analytics.usa.gov/data/live/realtime.json](https://analytics.usa.gov/data/live/realtime.json)

### 14. Write a Python program get the number of security alerts issued by US-CERT in the current year.

*Source:* [https://www.us-cert.gov/ncas/alerts](https://www.us-cert.gov/ncas/alerts)

### 15. Write a Python program to get the number of Pinterest accounts maintained by U.S. State Department embassies and missions

*Source:* [https://www.state.gov/r/pa/ode/socialmedia/](https://www.state.gov/r/pa/ode/socialmedia/)

### 16. Write a Python program to get the number of followers of a given twitter account.

### 17. Write a Python program to get the number of following on Twitter.

### 18. Write a Python program to get the number of post on Twitter liked by a given account

### 19. Write a Python program to count number of tweets by a given Twitter account.

### 20. Write a Python program to scrap number of tweets of a given Twitter account.

### 21. Write a Python program to find the live weather report (temperature, wind speed, description and weather) of a given city.

### 22. Write a Python program to display the date, days, title, city, country of next 25 Hackevents.

### 23. Write a Python program to download IMDB's Top 250 data (movie name, Initial release, director name and stars).

### 24. Write a Python program to get movie name, year and a brief summary of the top 10 random movies.

### 25. Write a Python program to get the number of magnitude 4.5+ earthquakes detected worldwide by the USGS.

### 26. Write a Python program to display the contains of different attributes like different attributes like status_code, headers, url, history, encoding, reason, cookies, elapsed, request and content of a specified resource.

### 27. Write a Python program to verifiy SSL certificates for HTTPS requests using requests module.

**Note:** Requests verifies SSL certificates for HTTPS requests, just like a web browser. By default, SSL verification is enabled, and Requests will throw a SSLError if it's unable to verify the certificate

#### Problem's Source:[https://www.w3resource.com/python-exercises/web-scraping/index.php](https://www.w3resource.com/python-exercises/web-scraping/index.php)